tcpdump/4.9.3/tcpdump-4.9.3.patch

```diff
@@ -0,0 +1,176 @@
++[
++	CPPFLAGS="$CPPFLAGS -I${withval}/include/ -I${withval}/include/wolfssl"
++	LDFLAGS="$LDFLAGS -L${withval}/lib"
++	LIBS="$LIBS -lwolfssl"
++	USE_WOLFSSL=yes],
++[
++	USE_WOLFSSL=no
++])
++
++if test $USE_WOLFSSL = yes
++then
++    AC_CHECK_HEADER([wolfssl/options.h])
++    if test $ac_cv_header_wolfssl_options_h = yes
++    then
++        AC_DEFINE(HAVE_WOLFSSL, 1, [define if you're using wolfSSL])
++        AC_DEFINE(HAVE_EVP_CIPHER_CTX_NEW, 1, [])
++        AC_DEFINE(HAVE_EVP_CIPHERINIT_EX, 1, [])
++        AC_DEFINE(HAVE_LIBCRYPTO, 1, [])
++        AC_DEFINE(HAVE_OPENSSL_EVP_H, 1, [])
++    else
++        AC_MSG_ERROR([Unable to find wolfssl.])
++    fi
++fi
++
+ # Check for OpenSSL/libressl libcrypto
+ AC_MSG_CHECKING(whether to use OpenSSL/libressl libcrypto)
+ # Specify location for both includes and libraries.
+@@ -872,7 +899,11 @@ AC_ARG_WITH(crypto,
+     AS_HELP_STRING([--with-crypto]@<:@=DIR@:>@,
+ 		   [use OpenSSL/libressl libcrypto (located in directory DIR, if specified) @<:@default=yes, if available@:>@]),
+ [
+-	if test $withval = no
++	if test $USE_WOLFSSL = yes
++	then
++		want_libcrypto=no
++		AC_MSG_RESULT(no)
++	elif test $withval = no
+ 	then
+ 		# User doesn't want to link with libcrypto.
+ 		want_libcrypto=no
+@@ -898,12 +929,18 @@ AC_ARG_WITH(crypto,
+ 		LIBS="-L$withval/lib $LIBS"
+ 	fi
+ ],[
++	if test $USE_WOLFSSL = yes
++	then
++		want_libcrypto=no
++		AC_MSG_RESULT(no)
+ 	#
+ 	# Use libcrypto if it's present, otherwise don't; no directory
+ 	# was specified.
+ 	#
+-	want_libcrypto=ifavailable
+-	AC_MSG_RESULT([yes, if available])
++	else
++		want_libcrypto=ifavailable
++		AC_MSG_RESULT([yes, if available])
++	fi
+ ])
+ if test "$want_libcrypto" != "no"; then
+ 	#
+diff --git a/print-esp.c b/print-esp.c
+index 6fabff1..7e5c7b5 100644
+--- a/print-esp.c
++++ b/print-esp.c
+@@ -37,6 +37,10 @@
+  * is the simplest way of handling the dependency.
+  */
+ #ifdef HAVE_LIBCRYPTO
++#ifdef HAVE_WOLFSSL
++#include <wolfssl/options.h>
++#include <openssl/ssl.h>
++#endif
+ #ifdef HAVE_OPENSSL_EVP_H
+ #include <openssl/evp.h>
+ #else
+@@ -259,7 +263,13 @@ int esp_print_decrypt_buffer_by_ikev2(netdissect_options *ndo,
+ 		EVP_CIPHER_CTX_free(ctx);
+ 		(*ndo->ndo_error)(ndo, "can't allocate memory for decryption buffer");
+ 	}
++#ifdef HAVE_WOLFSSL
++	/* wolfSSL expects the length passed in to include the padding, so we use
++	 * buffer_size here instead of len. */
++	EVP_Cipher(ctx, output_buffer, input_buffer, buffer_size);
++#else
+ 	EVP_Cipher(ctx, output_buffer, input_buffer, len);
++#endif
+ 	EVP_CIPHER_CTX_free(ctx);
+ 
+ 	/*
+@@ -356,6 +366,12 @@ static int
+ espprint_decode_encalgo(netdissect_options *ndo,
+ 			char *decode, struct sa_list *sa)
+ {
++#ifdef HAVE_WOLFSSL
++	/* wolfSSL doesn't support these cipher aliases, so we map them manually. */
++	const char TRIPLE_DES[] = "3des";
++	const char AES_256[] = "aes256";
++	const char AES_128[] = "aes128";
```

Comment:
I agree that adding aes256 and aes128 to `alias_tbl` in `wolfSSL_EVP_get_cipherbyname` makes sense. I'm less sure how to handle 3des. This isn't an alias supported by OpenSSL, hence tcpdump calling `EVP_add_cipher_alias(SN_des_ede3_cbc, "3des")` to add it themselves. We can either patch tcpdump (which is what I'm doing now) for this specific case, or we can add a `EVP_add_cipher_alias` compatibility function that allows the user to augment `alias_tbl`. That means `alias_tbl` needs to become a static member of evp.c rather than a function-local variable. I'm also not sure of the broader implications of calling `EVP_add_cipher_alias` in OpenSSL...so perhaps we add aes256 and aes128 to `wolfSSL_EVP_get_cipherbyname` and keep the patch for 3des?

Reply:
"3des" is recognized as a valid name in boringssl (look at `EVP_get_cipherbyname` in `openssl/boringssl/crypto/cipher/cipher.c`). I think adding it to the `alias_tbl` won't hurt. This way you won't need to implement `EVP_add_cipher_alias` since that would be a poor use of your time.

Reply:
Yeah, I think it's probably fine. I was just hesitant because OpenSSL (not boringssl) doesn't support it. At least the version I checked.